### PR TITLE
Logql Analyzer

### PR DESCRIPTION
**Pull request by vlad-diachenko**
**What this PR does / why we need it**:
It's API server that returns detailed information about execution LogQL expression against user's log line. This API is used on https://grafana.com/docs/loki/latest/logql/analyzer/ page
**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated
- [ ] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: pkg/logql/log/pipeline.go [hunk 1]
```diff
@@ -118,6 +119,85 @@ func NewPipeline(stages []Stage) Pipeline {
...
+}
+
+type streamPipelineAnalyzer struct {
+	origin       *pipeline
+	stagesCount  int
+	streamLabels labels.Labels
+}
+
+func NewPipelineAnalyzer(origin Pipeline, streamLabels labels.Labels) PipelineAnalyzer {
+	if o, ok := origin.(*pipeline); ok {
```
Comment (owen-d) [inline pkg/logql/log/pipeline.go:140]: type switching here will work, but has the tendency to break when we refactor or extend the code (such as renaming `type pipeline struct` or adding another pipeline variant.

Instead, we can add a new interface

```go
type PipelineStages interface {
  Pipeline
  Stages() []Stage
}
```

That way, we can do two things:
1) Make the `PipelineAnalyzer` work for all types that implement `PipelineStages`, including those we may add in the future
2) Let's us move the `PipelineAnalyzer` code into the `logqlanalyzer` pkg as it no longer depends on the unexported `pipeline` struct.

What do you think?

File: pkg/logqlanalyzer/analyzer.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+			LabelsBefore: mapAllToLabelsResponse(record.LabelsBefore),
+			LineAfter:    record.LineAfter,
+			LabelsAfter:  mapAllToLabelsResponse(record.LabelsAfter),
+			FilteredOut:  record.FilteredOut,
+		})
+	}
+	return LineResult{originLine, stageRecords}
+}
+
+func mapAllToLabelsResponse(labels labels.Labels) []Label {
```
Comment (owen-d) [inline pkg/logqlanalyzer/analyzer.go:77]: This can be replaced with `Labels.Copy()`

**Review by owen-d — COMMENTED** [review]
Looking great, left a few notes.

File: cmd/logql-analyzer/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+WORKDIR /src/loki
+RUN make clean && CGO_ENABLED=0 go build ./cmd/logql-analyzer/
+
+FROM alpine:3.15.4
+
+RUN apk add --no-cache ca-certificates
+
+COPY --from=build /src/loki/logql-analyzer /usr/bin/logql-analyzer
+
+ENTRYPOINT [ "/usr/bin/logql-analyzer", "--server.http-listen-port=3001" ]
```
Comment (jeschkies) [inline cmd/logql-analyzer/Dockerfile:13]: Should we rather pull the port from a config file?

File: pkg/logql/log/pipeline.go [hunk 4]
```diff
@@ -118,6 +119,85 @@ func NewPipeline(stages []Stage) Pipeline {
 	}
 }
 
+type PipelineAnalyzer interface {
+	AnalyzeLine(line string) []StageAnalysisRecord
+}
+type noopPipelineAnalyzer struct {
+	_ *pipeline
```
[the remark this answers is not in the document]
Comment (vlad-diachenko) [reply pkg/logql/log/pipeline.go:126]: removed

File: cmd/logql-analyzer/Dockerfile (hunk 3 above)
Comment (vlad-diachenko) [reply cmd/logql-analyzer/Dockerfile:13]: we do not use the config files, just a few args to define the port and log level....anyway this port here is not needed, moved it to `docker-compose` file.

File: pkg/logql/log/pipeline.go (hunk 1 above)
Comment (vlad-diachenko) [reply pkg/logql/log/pipeline.go:140]: done

**Review by owen-d — APPROVED** [review]
LGTM!

**Pull request merged**